src/openfermion/utils/_trotter_qasm_conversion.py

```diff
@@ -0,0 +1,331 @@
+#	Licensed under the Apache License, Version 2.0 (the "License");
+#	you may not use this file except in compliance with the License.
+#	You may obtain a copy of the License at
+#
+#		http://www.apache.org/licenses/LICENSE-2.0
+#
+#	Unless required by applicable law or agreed to in writing, software
+#	distributed under the License is distributed on an "AS IS" BASIS,
+#	WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#	See the License for the specific language governing permissions and
+#	limitations under the License.
+
+from openfermion.ops import QubitOperator
+import numpy as np
+
+'''
+Description:
+		Approximates the exponential of an operator composed of Pauli
+		matrices, by Trotterization. Outputs QASM format to a python
+		stream.
+
+		Change-of-basis gates:
+		H			: Z to X basis (and back)
+		Rx(pi/2)	: Z to Y basis ..... 
+		Rx(-pi/2) 	: Y to Z basis ..... 
+
+'''
+
+
+
+
+def third_order_trotter_helper(hamiltonian_dict,trotterized_list,op_a, op_b):
+	'''
+	This function recursively trotterizes a QubitOperator according
+		to the scheme:
+		e^(A+B) = e^(7/24 * A) e^(2/3 * B) e^(3/4 * A) e^(-1/24 * A) e^(B)
+
+	Used in trotterize() function.
+
+	Params: 
+		trotterized_list: list that the function populates with trotterized
+				QubitOperators
+		op_a: first term in hamiltonian as a QubitOperator
+		op_b: the rest of the terms in the hamiltonian as a list of
+				QubitOperator term dictionary keys
+
+	Returns: None
+	'''
+	trotterized_list.append(7.0/24.0 * op_a)
+	if len(op_b) == 1:
+		trotterized_list.append(2.0 / 3.0 * QubitOperator(op_b[0],
+			hamiltonian_dict[op_b[0]]))
+	else:
+		btemp = op_b[1:]
+		for i in btemp:
+			hamiltonian_dict[i] *= 2.0 / 3.0
+		atemp = QubitOperator(op_b[0],hamiltonian_dict[op_b[0]])
+		atemp *= 2.0 / 3.0
+		third_order_trotter_helper(trotterized_list, atemp, btemp)
+	trotterized_list.append(3.0/4.0 * op_a)
+
+	if op_b.terms.size == 1:
+		trotterized_list.append(-2.0/3.0 * QubitOperator(op_b[0],
+			hamiltonian_dict[op_b[0]]))
+	else:
+		btemp = op_b[1:]
+		for i in btemp:
+			hamiltonian_dict[i] *= -2.0 / 3.0
+		atemp = QubitOperator(op_b[0],hamiltonian_dict[op_b[0]])
+		atemp *= -2.0 / 3.0
+		third_order_trotter_helper(trotterized_list,atemp,btemp)
+	trotterized_list.append(-1.0 / 24.0 * op_a)
+
+	if op_b.terms.size == 1:
+		trotterized_list.append(QubitOperator(op_b[0],
+			hamiltonian_dict[op_b[0]]))
+	else:
+		btemp = op_b[1:]
+		atemp = QubitOperator(op_b[0],hamiltonian_dict[op_b[0]])
+		third_order_trotter_helper(trotterized_list, atemp, btemp)
+
+
+
+
+def trotterize(
```

Comment:
Follow style with function declarations, but I'll assume you get the picture from here.

Reply:
A rather superficial issue is that "trotterize" is way too ambiguous of a name. Keep in mind that "trotterize" is going to be something that is imported generically from the utils module. So is this a function that returns a qasm string (it is), or is it one that returns a unitary matrix corresponding to the trotter circuit, etc.? It needs to be clear from the name.

But I have much deeper concerns about this function and I do not think it should exist at all. A general design principle of OpenFermion is based on the idea that circuit can sort of be specified at the level of evolution under a sequence of local QubitOperators (or even local FermionOperators). For instance, Trotterization is a sequence of evolution under local QubitOperators. But you can also think about something like unitary coupled cluster as being evolution under a sequence of FermionOperator, etc. I believe we should work towards developing modules which output circuits very abstractly as a list specifying evolution under such local operators in this sort of topological fashion.

What we need from a PR like this then is a function which one can map over that list which converts each local operator to a QASM string. So I really think the first pull request of this sort should only apply to local QubitOperators. Perhaps from there we can add a function which takes an ordered list of single QubitOperators along with a Trotter order and then outputs a new list of those QubitOperators corresponding to Trotter evolution at the given order.

While not necessary for this PR, note that it also makes sense for this code to accept single FermionOperators. That is because the QubitOperators that come out of single FermionOperators have special structure which allows them to be implemented without Trotterization. That is not necessary functionality for now though.

Reply:
For sure--point(s) taken.

Regarding your 2nd paragraph, just want to point out that one of the functions Eric wrote does in fact output a list of QubitOperators--so even at the moment it isn't purely a QASM generator. I'll make that clearer though and make the role of that functionality more prominent.